qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -266,63 +251,198 @@ def run(self, dag):
         output_layout = Layout({dag.qubits[k]: v for (k, v) in layout_mapping})
         self.property_set["final_layout"] = output_layout
         if not self.fake_run:
-            for node_id in gate_order:
-                node = dag._multi_graph[node_id]
-                process_swaps(
-                    swap_map,
-                    node,
-                    mapped_dag,
-                    original_layout,
-                    canonical_register,
-                    self.fake_run,
-                    self._qubit_indices,
+            mapped_dag = dag.copy_empty_like()
+            _apply_sabre_result(
+                mapped_dag,
+                dag,
+                self._qubit_indices,
+                canonical_register,
+                original_layout,
+                sabre_result,
+                circuit_to_dag_dict,
+            )
+            return mapped_dag
+        return dag
+
+
+def _build_sabre_dag(dag, num_qubits, num_clbits, qubit_indices, clbit_indices):
+    from qiskit.converters import circuit_to_dag
+
+    # Maps id(block): circuit_to_dag(block) for all descendant blocks
+    circuit_to_dag_dict = {}
+
+    def recurse(block, block_qubit_indices, block_clbit_indices):
+        block_dag = circuit_to_dag(block)
+        circuit_to_dag_dict[id(block)] = block_dag
+        return process_dag(block_dag, block_qubit_indices, block_clbit_indices)
```

Comment:
Technically this `circuit_to_dag_dict` can have its keys overridden during the construction, like if we did something like this:
```python
import itertools
from qiskit import QuantumCircuit

inner = QuantumCircuit(2)
inner.cx(0, 1)

outer = QuantumCircuit(4, 1)
for pair in itertools.permutations(range(outer.num_qubits), 2):
    outer.if_test((outer.cregs[0], 1), inner, pair, [])
```

That said, I don't think it actually matters (except that we're doing extra conversion work that we don't need to be doing).  If we don't already have a test that such a circuit works though, it would be good to add one.

Reply:
Addressed in e74b157.